cmd/opm/registry/serve.go

```diff
@@ -66,92 +78,122 @@ func serveFunc(cmd *cobra.Command, args []string) error {
-	if err != nil {
-		logger.Fatalf("failed to listen: %s", err)
+	if storeErr != nil {
+		return fmt.Errorf("failed to load grpc store: %v", storeErr)
 	}
 
-	timeout, err := cmd.Flags().GetString("timeout-seconds")
+	lis, err := net.Listen("tcp", ":"+s.port)
 	if err != nil {
-		return err
+		s.logger.Fatalf("failed to listen: %s", err)
 	}
 
-	s := grpc.NewServer()
-	logger.Printf("Keeping server open for %s seconds", timeout)
-	if timeout != "infinite" {
-		timeoutSeconds, err := strconv.ParseUint(timeout, 10, 16)
+	grpcServer := grpc.NewServer()
+	s.logger.Printf("Keeping server open for %s seconds", s.timeout)
+	if s.timeout != "infinite" {
+		timeoutSeconds, err := strconv.ParseUint(s.timeout, 10, 16)
 		if err != nil {
 			return err
 		}
 
 		timeoutDuration := time.Duration(timeoutSeconds) * time.Second
 		timer := time.AfterFunc(timeoutDuration, func() {
-			logger.Info("Timeout expired. Gracefully stopping.")
-			s.GracefulStop()
+			s.logger.Info("Timeout expired. Gracefully stopping.")
+			grpcServer.GracefulStop()
 		})
 		defer timer.Stop()
 	}
 
-	api.RegisterRegistryServer(s, server.NewRegistryServer(store))
-	health.RegisterHealthServer(s, server.NewHealthServer())
-	reflection.Register(s)
-	logger.Info("serving registry")
-	return graceful.Shutdown(logger, func() error {
-		return s.Serve(lis)
+	api.RegisterRegistryServer(grpcServer, server.NewRegistryServer(store))
+	health.RegisterHealthServer(grpcServer, server.NewHealthServer())
+	reflection.Register(grpcServer)
+	s.logger.Info("serving registry")
+	return graceful.Shutdown(s.logger, func() error {
+		return grpcServer.Serve(lis)
 	}, func() {
-		s.GracefulStop()
+		grpcServer.GracefulStop()
 	})
 }
 
-func migrate(cmd *cobra.Command, db *sql.DB) error {
-	shouldSkipMigrate, err := cmd.Flags().GetBool("skip-migrate")
+func declcfgQuerier(cfg declcfg.DeclarativeConfig) (registry.GRPCQuery, error) {
+	m, err := declcfg.ConvertToModel(cfg)
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("could not build index model from declarative config: %v", err)
+	}
+	q := registry.NewQuerier(m)
+	return q, nil
+}
+
+func (s *serve) loadDBStore(ctx context.Context, source string) (registry.GRPCQuery, error) {
+	db, err := sqlite.Open(source)
+	if err != nil {
+		return nil, err
 	}
-	if shouldSkipMigrate {
-		return nil
+
+	if !s.skipMigrate {
+		// migrate to the latest version
+		if err := s.migrate(ctx, db); err != nil {
+			s.logger.WithError(err).Warnf("couldn't migrate db")
+		}
+	}
+
+	dbStore := sqlite.NewSQLLiteQuerierFromDb(db)
+
+	// sanity check that the db is available
+	tables, err := dbStore.ListTables(ctx)
+	if err != nil {
+		s.logger.WithError(err).Warnf("couldn't list tables in db")
 	}
+	if len(tables) == 0 {
+		s.logger.Warn("no tables found in db")
+	}
+
+	if s.skipMigrate {
+		return dbStore, nil
+	}
+
+	m, err := sqlite.ToModel(ctx, dbStore)
+	if err != nil {
+		return nil, err
+	}
+	return registry.NewQuerier(m), nil
```

Comment:
Do we want to serve sqlite databases from the model, or should we just return the `dbStore` directly (which would mean continuing to serve sqlite from the sqlite querier)?

There's some risk involved in serving sqlite from the model because that's a change from current code paths that has not been thoroughly tested.

I suppose if we started getting bugs against this, it would be simple to revert back to serving directly from sqlite.

Reply:
Plan discussed offline: by default keep serving sqlite databases as we always have.

If we want to load the sqlite db into memory and serve it from the model, we can do that really easily in a follow-up PR (probably gated by a feature flag).